content/ADR-164-quests.poc.md

```diff
@@ -0,0 +1,211 @@
+
+Users may ask for quests state and metrics, but completed Quests may be stored for a certain time if the system needs space or it may grow undefinetly.
+
+- The system should be able to process all events in parallel, in order to make progress in corresponding Quests.
+- The system should be able to accept connections and subscriptions to Quest instances.
+- The system should be able to send updates to subscribed users.
+- The system should be able to respond to Quests state requests.
+- The system should expose metrics related to the Quests.
+
+### Workflows: Quest User
+
+- Start Quest
+
+There would be several ways to discovery a Quest. The user may be interested to start any of them. 
+
+```mermaid
+sequenceDiagram
+    Quest Discovery-->>Explorer/Kernel: New quest available
+    Explorer/Kernel->>Quest Server: Send start quest request 
+    Quest Server ->> Explorer/Kernel: Send start quest response
+    loop
+        Quest Server -->> Explorer/Kernel: Send quest updates
+    end
+```
+
+- Track Progress 
+
+Client should make progress by sending verified events
+```mermaid
+sequenceDiagram
+    Explorer/Kernel ->> Catalyst: Ask for signed event 
+    Explorer/Kernel ->> Quest Server: Send signed event
+    loop
+        Quest Server -->> Explorer/Kernel: Send quest updates
+    end
+```
+
+- Real time updates
+
+Once the API is defined, the system may expose different ways to query quest states, one of them is to use some long living channel (let's say WebSockets).
+
+```mermaid
+sequenceDiagram
+    Explorer/Kernel ->> Quest Server: Send subscription request
+    loop
+        Quest Server -->> Explorer/Kernel: Send quest updates
+    end
+```
+
+### Workflows: Quest Designer 
+
+- Deploy new quest or modify existing quest
+
+When the quest definition is ready, the Quest Designer tool would be able to send the new or modified version to the Quest Server. The request should include the author signature.
+
+```mermaid
+sequenceDiagram
+    Quest Designer ->> Quest Designer: Quest Definition 
+    Quest Designer ->> Quest Server: Send Quest definition and signature
+    Quest Server ->> Quest Designer: Send Quest submission response 
+```
+
+- Retrieve quest stats
+
+The quest author may be interested in the stats, how many users started, completed, and other information like starting and ending times.
+
+```mermaid
+sequenceDiagram
+    Quest Designer ->> Quest Server: Query Quest state (include signature to validate authorship)
+    Quest Server ->> Quest Designer: Send Quest state 
+```
+
+## Specification
+
+About the implementation and alternatives, each actor has some important decisions to explore.
+
+### Communication between the System and Kernel
+
+In order to receive real time updates, there are two options to handle the communication:
+- Use a socket (peek any protocol like WebSocket, WebRTC, WebTransport, QUIC, etc) to keep posted on interested quests. 
```

Comment:
Is this more related to _two way communications_ (i.e. being able to push) or having a long-lived communication channel?

Reply:
It's more related to long-lived communication and async event processing.